bundle/edu.gemini.shared.util/src/main/java/edu/gemini/shared/util/immutable/DefaultImList.java

```diff
@@ -348,6 +348,21 @@ public T max(final Comparator<? super T> c) {
     }
 
     @Override
+    public <K> HashMap<K, ImList<T>> groupBy(Function1<? super T, K> f) {
+        HashMap<K, ImList<T>> m = new HashMap<>();
+        for (final T elem : backingList) {
+            final K key = f.apply(elem);
+            final ImList<T> val = m.get(key);
+            if (val == null) {
+                m.put(key, create(elem));
+            } else {
+                m.put(key, val.append(elem));
+            }
```

Comment:
I think this can be:

```
m.put(key, ImOption.apply(m.get(key)).map(lst-> lst.append(elem)).getOrElse(() -> create(elem));
```

Though maybe not any better.

Reply:
Actually, this made realize `ImOption.apply(null)` returns `None`, so I will use this technique to handle all those nulls when the key is not present.